File: onhire_pro/report/rental_item_availability_calendar/rental_item_availability_calendar.py
# Project/onhire_pro/onhire_pro/report/rental_item_availability_calendar/rental_item_availability_calendar.py
import frappe
from frappe import _
from frappe.utils import getdate, nowdate, add_days, get_datetime_str

def execute(filters=None):
    columns = get_columns(filters)
    data = get_data(filters)
    # Chart and summary are not typically used when rendering a custom calendar like this in a script report
    return columns, data, None, None, None 

def get_columns(filters):
    # These columns are for the underlying data table, which might be hidden or less prominent
    # if the main focus is the calendar visualization injected by JS.
    return [
        {
            "label": _("Item Code"),
            "fieldname": "item_code",
            "fieldtype": "Link",
            "options": "Item",
            "width": 120
        },
        {
            "label": _("Serial No"),
            "fieldname": "serial_no",
            "fieldtype": "Link",
            "options": "Serial No",
            "width": 120
        },
        {
            "label": _("Event Type"), # 'Rental Job' or 'Maintenance Task'
            "fieldname": "event_type",
            "fieldtype": "Data",
            "width": 120
        },
        {
            "label": _("Reference Doctype"),
            "fieldname": "reference_doctype",
            "fieldtype": "Data",
            "hidden": 1
        },
        {
            "label": _("Reference Name"),
            "fieldname": "reference_name",
            "fieldtype": "Dynamic Link",
            "options": "reference_doctype",
            "width": 150
        },
        {
            "label": _("Status"),
            "fieldname": "status",
            "fieldtype": "Data",
            "width": 100
        },
        {
            "label": _("Start Date"),
            "fieldname": "start_date",
            "fieldtype": "Datetime",
            "width": 150
        },
        {
            "label": _("End Date"),
            "fieldname": "end_date",
            "fieldtype": "Datetime",
            "width": 150
        },
        {
            "label": _("Customer"),
            "fieldname": "customer",
            "fieldtype": "Link",
            "options": "Customer",
            "width": 150
        }
    ]

def get_data(filters):
    data = []
    company = filters.get("company")
    from_date = filters.get("from_date")
    to_date = filters.get("to_date")
    item_code_filter = filters.get("item_code")
    item_group_filter = filters.get("item_group")
    event_type_filter = filters.get("status") # Filter name is 'status' in JS, maps to event_type

    sql_params = {
        "company": company,
        "from_date": from_date,
        "to_date": to_date
    }

    # Fetch Rental Jobs
    if not event_type_filter or event_type_filter == "Rental Job":
        rj_conditions = ""
        current_sql_params_rj = sql_params.copy()
        if item_code_filter:
            rj_conditions += " AND rji.item_code = %(item_code)s"
            current_sql_params_rj["item_code"] = item_code_filter
        if item_group_filter:
            rj_conditions += " AND item.item_group = %(item_group)s"
            current_sql_params_rj["item_group"] = item_group_filter
        
        rental_jobs_query = f"""
            SELECT 
                rji.item_code, 
                rji.serial_no,
                'Rental Job' as event_type,
                rj.name as reference_name,
                'Rental Job' as reference_doctype,
                rj.status,
                rj.scheduled_dispatch_date as start_date,
                rj.scheduled_return_date as end_date,
                rj.customer
            FROM `tabRental Job Item` rji
            JOIN `tabRental Job` rj ON rji.parent = rj.name
            {'JOIN `tabItem` item ON rji.item_code = item.name' if item_group_filter else ''}
            WHERE rj.docstatus = 1
              AND rj.company = %(company)s
              AND rj.status NOT IN ('Cancelled', 'Draft')
              AND rj.scheduled_dispatch_date <= %(to_date)s
              AND rj.scheduled_return_date >= %(from_date)s
              {rj_conditions}
        """
        rental_job_data = frappe.db.sql(rental_jobs_query, current_sql_params_rj, as_dict=True)
        data.extend(rental_job_data)

    # Fetch Maintenance Tasks
    if (not event_type_filter or event_type_filter == "Maintenance Task") and frappe.db.exists("DocType", "Maintenance Task"):
        mt_conditions = ""
        current_sql_params_mt = sql_params.copy()

        # Check for company field in Maintenance Task
        mt_company_filter = ""
        if frappe.db.has_column("Maintenance Task", "company"):
            mt_company_filter = "AND mt.company = %(company)s"
        else:
            if "company" in current_sql_params_mt: # remove if not applicable
                 del current_sql_params_mt["company"]


        if item_code_filter and frappe.db.has_column("Maintenance Task", "item_code"):
            mt_conditions += " AND mt.item_code = %(item_code)s"
            current_sql_params_mt["item_code"] = item_code_filter
        
        item_join_mt = ""
        if item_group_filter and frappe.db.has_column("Maintenance Task", "item_code"):
            item_join_mt = " JOIN `tabItem` item ON mt.item_code = item.name "
            mt_conditions += " AND item.item_group = %(item_group)s"
            current_sql_params_mt["item_group"] = item_group_filter

        # Determine date fields for Maintenance Task
        start_date_field_mt = "mt.start_date" if frappe.db.has_column("Maintenance Task", "start_date") else "mt.creation"
        end_date_field_mt = "mt.completion_date"
        if not frappe.db.has_column("Maintenance Task", "completion_date") and frappe.db.has_column("Maintenance Task", "expected_completion_date"):
            end_date_field_mt = "mt.expected_completion_date"
        elif not frappe.db.has_column("Maintenance Task", "completion_date"):
             # If neither completion_date nor expected_completion_date exists, this query part might fail or return nulls for end_date
             # For simplicity, we'll proceed assuming one of them exists, or the query will just not find matches for date range on end_date
             # A robust solution would handle this by perhaps not filtering on end_date if the field is missing.
             pass


        maintenance_tasks_query = f"""
            SELECT 
                mt.item_code, 
                mt.serial_no,
                'Maintenance Task' as event_type,
                mt.name as reference_name,
                'Maintenance Task' as reference_doctype,
                mt.status,
                {start_date_field_mt} as start_date,
                {end_date_field_mt} as end_date, 
                NULL as customer
            FROM `tabMaintenance Task` mt
            {item_join_mt}
            WHERE mt.docstatus = 1
              {mt_company_filter}
              AND mt.status NOT IN ('Cancelled', 'Completed') 
              AND {start_date_field_mt} <= %(to_date)s
              AND {end_date_field_mt} >= %(from_date)s
              {mt_conditions}
        """
        if frappe.db.has_column("Maintenance Task", "start_date") and \
           (frappe.db.has_column("Maintenance Task", "completion_date") or frappe.db.has_column("Maintenance Task", "expected_completion_date")):
            maintenance_data = frappe.db.sql(maintenance_tasks_query, current_sql_params_mt, as_dict=True)
            data.extend(maintenance_data)
            
    return data

File: onhire_pro/report/rental_item_availability_calendar/rental_item_availability_calendar.js
// Project/onhire_pro/onhire_pro/report/rental_item_availability_calendar/rental_item_availability_calendar.js
frappe.query_reports["Rental Item Availability Calendar"] = {
    "filters": [
        {
            "fieldname": "company",
            "label": __("Company"),
            "fieldtype": "Link",
            "options": "Company",
            "default": frappe.defaults.get_user_default("Company"),
            "reqd": 1
        },
        {
            "fieldname": "from_date",
            "label": __("From Date"),
            "fieldtype": "Date",
            "default": frappe.datetime.now_date(),
            "reqd": 1
        },
        {
            "fieldname": "to_date",
            "label": __("To Date"),
            "fieldtype": "Date",
            "default": frappe.datetime.add_days(frappe.datetime.now_date(), 30),
            "reqd": 1
        },
        {
            "fieldname": "item_code",
            "label": __("Item Code"),
            "fieldtype": "Link",
            "options": "Item",
            "get_query": function() {
                return { filters: { "is_rental_item": 1 } }
            }
        },
        {
            "fieldname": "item_group",
            "label": __("Item Group"),
            "fieldtype": "Link",
            "options": "Item Group"
        },
        {
            "fieldname": "status",
            "label": __("Event Type"),
            "fieldtype": "Select",
            "options": "\nRental Job\nMaintenance Task", // Add more if needed
            "default": "Rental Job"
        }
    ],
    "formatter": function(value, row, column, data, Rreport) {
        if (column.id === "item_code" && value) {
            return `<a href="/app/item/${value}">${value}</a>`;
        }
        if (column.id === "reference_name" && data.reference_doctype) {
             return `<a href="/app/${data.reference_doctype.toLowerCase().replace(/ /g,'-')}/${value}">${value}</a>`;
        }
        return value;
    },
    "initial_depth": 1, // For tree view if using resources
    "onload": function(report) {
        // Custom rendering logic for calendar/timeline will go here
        // This report type might be better as a Page with a custom HTML field for FullCalendar
        // For a Script Report, we'd typically generate columns/data for a table.
        // To show a calendar, we'd need to inject HTML and JS.
        
        report.page.add_field({
            fieldname: 'calendar_view_html',
            fieldtype: 'HTML',
            label: 'Calendar View'
        });
        
        report.render_calendar = function(data) {
            if (!data) data = report.data; // Use report data if not passed
            var $calendar_area = $(report.page.fields_dict.calendar_view_html.wrapper).empty();
            $calendar_area.css({height: '600px', marginBottom: '20px'}); // Ensure calendar has height
            
            if (!data.length && !report.loading) {
                $calendar_area.html(`<p class="text-muted">${__("No data to display in calendar.")}</p>`);
                return;
            }

            // Ensure FullCalendar is loaded
            if (typeof FullCalendar === 'undefined') {
                $calendar_area.html(`<p class="text-danger">${__("FullCalendar library not loaded.")}</p>`);
                return;
            }
            
            let calendarEl = $calendar_area.get(0);

            let calendar = new FullCalendar.Calendar(calendarEl, {
                headerToolbar: {
                    left: 'prev,next today',
                    center: 'title',
                    right: 'dayGridMonth,timeGridWeek,timeGridDay,listWeek'
                },
                initialDate: report.filters.from_date || frappe.datetime.now_date(),
                editable: false,
                eventLimit: true, // allow "more" link when too many events
                events: data.map(function(item) {
                    let title = item.item_code || 'Event';
                    if(item.serial_no) title += ` (${item.serial_no})`;
                    if(item.reference_name) title += ` - ${item.reference_name}`;
                    title += ` (${item.status || item.event_type})`;
                    
                    let color = '#3a87ad'; // Default
                    if(item.event_type === 'Rental Job') {
                        color = item.status === 'Completed' ? '#d3d3d3' : (item.status === 'Cancelled' ? '#f64c72' : '#28a745');
                    } else if (item.event_type === 'Maintenance Task') {
                        color = item.status === 'Completed' ? '#6c757d' : '#ffc107'; // Yellow for maintenance
                    }

                    return {
                        title: title,
                        start: item.start_date, 
                        end: frappe.datetime.add_days(item.end_date, 1), 
                        allDay: true, 
                        color: color,
                        url: item.reference_doctype && item.reference_name ? frappe.utils.get_form_link(item.reference_doctype, item.reference_name) : null
                    };
                }),
                eventDidMount: function(info) {
                    $(info.el).tooltip({ title: info.event.title });
                }
            });
            calendar.render();
            report.calendar_instance = calendar; // Store instance if needed
        };

        // Override the report render method to include calendar rendering
        let original_render = report.render;
        report.render = function(data) {
            original_render.call(report, data); 
            report.render_calendar(data); 
        };
        
        if(report.data && report.data.length > 0){
            report.render_calendar(report.data);
        } else {
            // Ensure calendar area is cleared if no data initially
             $(report.page.fields_dict.calendar_view_html.wrapper).empty().html(`<p class="text-muted">${__("Apply filters to load calendar data.")}</p>`);
        }
    }
};
